engine/swarm/secrets.md

```diff
@@ -0,0 +1,812 @@
+      `mysql` password in the bind-mounted `tempdir` volume. The `mysql` image
+      reads the password string from that file when initializing the database
+      for the first time. Afterward, it is stored in the MySQL database itself.
+
+      ```bash
+      $ docker service create \
+                       --name mysql \
+                       --replicas 1 \
+                       --network mysql_private \
+                       --mount type=volume,source=mydata,destination=/var/lib/mysql \
+                       --secret source=mysql_password,target=mysql_password \
+                       -e MYSQL_ROOT_PASSWORD_FILE="/run/secrets/mysql_password" \
+                      mysql:latest
+
+      i15rvmizjsjkn02nls214dluk
+      ```
+
+3.  Verify that the `mysql` container is running using the `docker service ls` command.
+
+    ```bash
+    $ docker service ls
+    ID            NAME   MODE        REPLICAS  IMAGE
+    i15rvmizjsjk  mysql  replicated  1/1       mysql:latest
+    ```
+
+    At this point, you could actually revoke the `mysql` service's access to the
+    `mysql_password` secret because the password has been saved in the MySQL
+    system database. Don't do that for now, because we will use it later to
+    facilitate rotating the MySQL password.
+
+4.  Now that MySQL is set up, create a WordPress service that connects to the
+    MySQL service. The WordPress service has the following characteristics:
+
+    - Because the scale is set to `1`, only a single WordPress task runs.
+      Load-balancing WordPress is left as an exercise to the reader, because of
+      limitations with storing WordPress session data on the container
+      filesystem.
+    - Exposes WordPress on port 30000 of the host machine, so that you can access
+      it from external hosts. You can expose port 80 instead if you do not have
+      a web server running on port 80 of the host machine.
+    - Connects to the `mysql_private` network so it can communicate with the
+      `mysql` container, and also publishes port 80 to port 30000 on all swarm
+      nodes.
+    - Has access to the `mysql_password` secret, but specifies a different
+      target file name within the container. The WordPress container will use
+      the mount point `/run/secrets/wp_db_password`. Also specifies that the
+      secret is not group-or-world-readable, by setting the mode to
+      `0400`.
+    - Sets the environment variable `WORDPRESS_DB_PASSWORD_FILE` to the file
+      path where the secret is mounted. The WordPress service will read the
+      MySQL password string from that file and add it to the `wp-config.php`
+      configuration file.
+
+    ```bash
+    $ docker service create \
+                     --name wordpress \
+                     --replicas 1 \
+                     --network mysql_private \
+                     --publish 30000:80 \
+                     --secret source=mysql_password,target=wp_db_password,mode=0400 \
```

Comment:
Add a volume here, so that the WordPress data is preserved when updating the service;

```
--mount type=volume,source=wpdata,destination=/var/www/html \
```